designs/drift-hash-versioning.md

```diff
@@ -0,0 +1,111 @@
+
+1. Setting cloud provider defaults for the: `nodeClassRef.apiVersion` and `nodeClassRef.kind` fields [#909](https://github.com/kubernetes-sigs/karpenter/issues/909)
+ -  **We will need to coordinate with the Azure team to make sure that `Kind` and `apiVersion` are also set on the (karpenter-provider-azure)[https://github.com/Azure/karpenter-provider-azure]**
+2. Fixing `volumeSize` to be correctly considered for hashing [#5447](https://github.com/aws/karpenter-provider-aws/issues/5447)
+
+More generally, these are breaking changes to the hashing function that should not induce drift:
+
+1. A `NodePool` adds default values to an existing field
+2. A `NodePool` field is added with a default
+3. A `NodePool` field is removed
+
+## Solutions
+
+#### Example Spec for Option 1 and Option 2
+```
+apiVersion: karpenter.sh/v1beta1
+kind: NodePool
+metadata:
+  name: default
+  annotations: 
+    karpenter.sh/nodepool-hash: abcdef
+    karpenter.sh/nodepool-hash-version: v1
+    ...
+---
+apiVersion: karpenter.sh/v1beta1
+kind: NodeClaim 
+metadata:
+  name: default
+  annotations: 
+    karpenter.sh/nodepool-hash: abcdef
+    karpenter.sh/nodepool-hash-version: v1
+    ...
+```
+
+### Option 1: Add a hash version annotation and fallback to rehashing on the NodeClaim 
+
+As the Karpenter CRDs change, Karpenter’s underlying hashing function may also change. To handle the upgrade path, Karpenter will add a `nodepool-hash-version` into the karpenter controller. The `nodepool-hash-version` will be bumped when a breaking change is made to the hash function. The version bump will be done by the maintainers.
+
+ When Karpenter is spun up, the controller will check the `nodepool-hash-version`, and if needed, re-calculate the `NodePool` hash propagate the annotation down to the `NodeClaim`. Karpenter will assert that this `nodepool-hash-version` annotation is equivalent on both the `NodePool` and `NodeClaim` before doing a drift evaluation. 
+
+As a result, two annotations will be used analyze and orchestrate drift: 
+
+* nodepool-hash: (hash of the nodeclaim template on the nodepool)
+* nodepool-hash-version: (the NodePool hashing version defined by Karpenter)
+
+Option 1 will have two case for drift:
+
+1. When the CRD `nodepool-hash-version` matches between the `NodePool` and `NodeClaim`:
+    1. Drift will work as normal and the `NodePool` hash will be compared to evaluate for drift 
+2. When the CRD `nodepool-hash-version` does not match between the `NodePool` and `NodeClaim`:
+    1. Karpenter would recalculate the hash of the `NodePool` and back-propagate the updated value for the `karpenter.sh/nodepool-hash` annotation to the NodeClaims. Any `NodeClaims` that are already considered drifted will remain drifted if the `karpenter.sh/nodepool-hash-version` doesn't match
```

Comment:
If we're saying that nodeclaims that are already drifted will remain drifted if the hash version doesn't match, does that mean we don't need to recalculate the hash?

Reply:
Yeah, for those we will not update the hash

Reply:
I think @njtran is saying that you don't need to update either of them, it doesn't really matter what you do. You can just leave the NodeClaim around entirely without doing any updates.

Reply:
We will need to disable drift checking if the NodePool versions don't match, so we will need make to update the NodePool version and not update the hash for drifted NodeClaims

Reply:
> not update the hash for drifted NodeClaims

Got it, but you still want to update the version? I don't really think that it makes a major difference either way, I guess to some degree it's nice to have the drifted NodeClaims to be aligned with everything else on version

Reply:
Yeah, that was my thinking